File: api/src/main/java/com/lightbend/lagom/javadsl/api/Descriptor.java
/*
 * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
 */
package com.lightbend.lagom.javadsl.api;

import java.util.Optional;

import com.lightbend.lagom.javadsl.api.deser.*;
import com.lightbend.lagom.javadsl.api.security.UserAgentServiceIdentificationStrategy;
import com.lightbend.lagom.javadsl.api.transport.HeaderTransformer;
import com.lightbend.lagom.javadsl.api.transport.Method;
import com.lightbend.lagom.javadsl.api.transport.PathVersionedProtocolNegotiationStrategy;
import org.pcollections.*;
import java.lang.reflect.Type;
import java.util.Arrays;

/**
 * Describes a service.
 *
 * A descriptor is a set of calls descriptors that the service provides, coupled with metadata about how the
 * service and its calls are to be served.  Metadata may include versioning and migrations, SLA's, sharding
 * hints, circuit breaker strategies etc.
 */
public final class Descriptor {

    /**
     * Holds the service call itself.
     *
     * The implementations of this are intentionally opaque, as the mechanics of how the service call implementation
     * gets passed around is internal to Lagom.
     */
    public interface ServiceCallHolder {
    }

    /**
     * Describes a service call.
     */
    public static final class Call<Request, Response> {
        private final CallId callId;
        private final ServiceCallHolder serviceCallHolder;
        private final MessageSerializer<Request, ?> requestSerializer;
        private final MessageSerializer<Response, ?> responseSerializer;
        private final Optional<CircuitBreakerId> circuitBreaker;
        private final Optional<Boolean> autoAcl;


        Call(CallId callId, ServiceCallHolder serviceCallHolder,
             MessageSerializer<Request, ?> requestSerializer,
             MessageSerializer<Response, ?> responseSerializer, Optional<CircuitBreakerId> circuitBreaker,
             Optional<Boolean> autoAcl) {

            this.callId = callId;
            this.serviceCallHolder = serviceCallHolder;
            this.requestSerializer = requestSerializer;
            this.responseSerializer = responseSerializer;
            this.circuitBreaker = circuitBreaker;
            this.autoAcl = autoAcl;
        }

        /**
         * Get the id for the call.
         *
         * @return The id.
         */
        public CallId callId() {
            return callId;
        }

        /**
         * A holder for the service call.
         *
         * This holds a reference to the service call, in an implementation specific way.
         *
         * @return The service call holder.
         */
        public ServiceCallHolder serviceCallHolder() {
            return serviceCallHolder;
        }

        /**
         * Get the request message serializer.
         *
         * @return The request serializer.
         */
        public MessageSerializer<Request, ?> requestSerializer() {
            return requestSerializer;
        }

        /**
         * Get the response message serializer.
         *
         * @return The response serializer.
         */
        public MessageSerializer<Response, ?> responseSerializer() {
            return responseSerializer;
        }

        /**
         * Get the circuit breaker identifier.
         *
         * @return The circuit breaker identifier.
         */
        public Optional<CircuitBreakerId> circuitBreaker() {
          return circuitBreaker;
        }
    
        /**
         * Whether this service call should automatically define an ACL for the router to route external calls to it.
         *
         * @return Some value if this service call explicitly decides that it should have an auto ACL defined for it,
         *         otherwise empty.
         */
        public Optional<Boolean> autoAcl() {
            return autoAcl;
        }

        /**
         * Return a copy of this call descriptor with the given service call ID configured.
         *
         * @param callId
         *          The call id.
         * @return A copy of this call descriptor.
         */
        public Call<Request, Response> with(CallId callId) {
            return new Call<>(callId, serviceCallHolder, requestSerializer, responseSerializer, circuitBreaker,
                    autoAcl);
        }

        /**
         * Return a copy of this call descriptor with the given service call holder configured.
         *
         * @param serviceCallHolder
         *          The service call holder.
         * @return A copy of this call descriptor.
         */
        public Call<Request, Response> with(ServiceCallHolder serviceCallHolder) {
            return new Call<>(callId, serviceCallHolder, requestSerializer, responseSerializer, circuitBreaker,
                    autoAcl);
        }
    
        /**
         * Return a copy of this call descriptor with the given request message
         * serializer configured.
         *
         * @param requestSerializer
         *          The request serializer.
         * @return A copy of this call descriptor.
         */
        public Call<Request, Response> withRequestSerializer(MessageSerializer<Request, ?> requestSerializer) {
            return new Call<>(callId, serviceCallHolder, requestSerializer, responseSerializer, circuitBreaker,
                    autoAcl);
        }
    
        /**
         * Return a copy of this call descriptor with the given response message
         * serializer configured.
         *
         * @param responseSerializer
         *          The response serializer.
         * @return A copy of this call descriptor.
         */
        public Call<Request, Response> withResponseSerializer(MessageSerializer<Response, ?> responseSerializer) {
            return new Call<>(callId, serviceCallHolder, requestSerializer, responseSerializer, circuitBreaker,
                    autoAcl);
        }
    
        /**
         * Return a copy of this call descriptor with the given service call
         * configured.
         *
         * @param serviceCall
         *          The service call.
         * @return A copy of this call descriptor.
         */
        public Call<Request, Response> with(ServiceCall<Request, Response> serviceCall) {
            return new Call<>(callId, serviceCallHolder, requestSerializer, responseSerializer, circuitBreaker,
                    autoAcl);
        }
    
        /**
         * Return a copy of this call descriptor with the given circuit breaker
         * identifier configured.
         *
         * @param breakerId
         *          The configuration id of the circuit breaker
         * @return A copy of this call descriptor.
         */
        public Call<Request, Response> withCircuitBreaker(CircuitBreakerId breakerId) {
            return new Call<>(callId, serviceCallHolder, requestSerializer, responseSerializer,
                    Optional.ofNullable(breakerId), autoAcl);
        }

        /**
         * Return a copy of this call descriptor with autoAcl configured.
         *
         * @param autoAcl Whether an ACL will automatically be generated for the gateway to route calls to this service.
         * @return A copy of this call descriptor.
         */
        public Call<Request, Response> withAutoAcl(boolean autoAcl) {
            return new Call<>(callId, serviceCallHolder, requestSerializer, responseSerializer, circuitBreaker,
                    Optional.of(autoAcl));
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof Call)) return false;

            Call<?, ?> call = (Call<?, ?>) o;

            return callId.equals(call.callId);

        }

        @Override
        public int hashCode() {
            return callId.hashCode();
        }

        @Override
        public String toString() {
            return "Call{" +
                    "callId=" + callId +
                    ", serviceCallHolder=" + serviceCallHolder +
                    ", requestSerializer=" + requestSerializer +
                    ", responseSerializer=" + responseSerializer +
                    ", circuitBreaker=" + circuitBreaker +
                    ", autoAcl=" + autoAcl +
                    '}';
        }
    }

    /**
     * The Call ID.
     *
     * This is an abstract representation of how a service call is addressed within a service.  For example, in the
     * case of REST APIs, it will be addressed using an HTTP method and path.
     */
    public static abstract class CallId {
        private CallId() {}
    }

    /**
     * A REST call ID.
     */
    public static final class RestCallId extends CallId {
        private final Method method;
        private final String pathPattern;

        public RestCallId(Method method, String pathPattern) {
            this.method = method;
            this.pathPattern = pathPattern;
        }

        /**
         * The HTTP method for the call.
         *
         * The method will only be used for strict REST calls. For other calls, such as calls implemented by WebSockets
         * or other transports, the method may be ignored completely.
         *
         * @return The HTTP method.
         */
        public Method method() {
            return method;
        }

        /**
         * The path pattern for the call.
         *
         * @return The path pattern.
         */
        public String pathPattern() {
            return pathPattern;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof RestCallId)) return false;

            RestCallId that = (RestCallId) o;

            if (!method.equals(that.method)) return false;
            return pathPattern.equals(that.pathPattern);

        }

        @Override
        public int hashCode() {
            int result = method.hashCode();
            result = 31 * result + pathPattern.hashCode();
            return result;
        }

        @Override
        public String toString() {
            return "RestCallId{" +
                    "method=" + method +
                    ", pathPattern='" + pathPattern + '\'' +
                    '}';
        }
    }

    /**
     * A path based call ID.
     */
    public static class PathCallId extends CallId {
        private final String pathPattern;

        public PathCallId(String pathPattern) {
            this.pathPattern = pathPattern;
        }

        /**
         * Get the path pattern.
         *
         * @return The path pattern.
         */
        public String pathPattern() {
            return pathPattern;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof PathCallId)) return false;

            PathCallId that = (PathCallId) o;

            return pathPattern.equals(that.pathPattern);

        }

        @Override
        public int hashCode() {
            return pathPattern.hashCode();
        }

        @Override
        public String toString() {
            return "PathCallId{" +
                    "pathPattern='" + pathPattern + '\'' +
                    '}';
        }
    }

    /**
     * A named call ID.
     */
    public static class NamedCallId extends CallId {
        private final String name;

        public NamedCallId(String name) {
            this.name = name;
        }

        /**
         * Get the name of the call.
         *
         * @return The name of the call.
         */
        public String name() {
            return name;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof NamedCallId)) return false;

            NamedCallId that = (NamedCallId) o;

            return name.equals(that.name);

        }

        @Override
        public int hashCode() {
            return name.hashCode();
        }

        @Override
        public String toString() {
            return "NamedCallId{" +
                    "name='" + name + '\'' +
                    '}';
        }
    }
    
    public static class CircuitBreakerId {
        private final String id;
  
        public CircuitBreakerId(String id) {
            this.id = id;
        }
  
        /**
         * Get the identifier of the circuit breaker
         */
        public String id() {
            return id;
        }
  
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (!(o instanceof CircuitBreakerId)) return false;
  
            CircuitBreakerId that = (CircuitBreakerId) o;
            return id.equals(that.id);
        }
  
        @Override
        public int hashCode() {
            return id.hashCode();
        }
  
        @Override
        public String toString() {
            return "CircuitBreakerId{" +
                    "id='" + id + '\'' +
                    '}';
        }
    }

    private final String name;
    private final PSequence<Call<?, ?>> calls;
    private final PMap<Type, PathParamSerializer<?>> pathParamSerializers;
    private final PMap<Type, MessageSerializer<?, ?>> messageSerializers;
    private final SerializerFactory serializerFactory;
    private final ExceptionSerializer exceptionSerializer;
    private final boolean autoAcl;
    private final PSequence<ServiceAcl> acls;
    private final HeaderTransformer protocolNegotiationStrategy;
    private final HeaderTransformer serviceIdentificationStrategy;
    private final boolean locatableService;

    Descriptor(String name) {
        this(name, TreePVector.empty(), HashTreePMap.empty(), HashTreePMap.empty(), SerializerFactory.DEFAULT,
                ExceptionSerializer.DEFAULT, false, TreePVector.empty(), new PathVersionedProtocolNegotiationStrategy(),
                new UserAgentServiceIdentificationStrategy(), true);
    }

    Descriptor(String name, PSequence<Call<?, ?>> calls, PMap<Type, PathParamSerializer<?>> pathParamSerializers,
            PMap<Type, MessageSerializer<?, ?>> messageSerializers, SerializerFactory serializerFactory,
            ExceptionSerializer exceptionSerializer, boolean autoAcl, PSequence<ServiceAcl> acls, HeaderTransformer protocolNegotiationStrategy,
            HeaderTransformer serviceIdentificationStrategy, boolean locatableService) {
        this.name = name;
        this.calls = calls;
        this.pathParamSerializers = pathParamSerializers;
        this.messageSerializers = messageSerializers;
        this.serializerFactory = serializerFactory;
        this.exceptionSerializer = exceptionSerializer;
        this.autoAcl = autoAcl;
        this.acls = acls;
        this.protocolNegotiationStrategy = protocolNegotiationStrategy;
        this.serviceIdentificationStrategy = serviceIdentificationStrategy;
        this.locatableService = locatableService;
    }

    public String name() {
        return name;
    }

    public PSequence<Call<?, ?>> calls() {
        return calls;
    }

    public PMap<Type, PathParamSerializer<?>> pathParamSerializers() {
        return pathParamSerializers;
    }

    public PMap<Type, MessageSerializer<?, ?>> messageSerializers() {
        return messageSerializers;
    }

    public SerializerFactory serializerFactory() {
        return serializerFactory;
    }

    public ExceptionSerializer exceptionSerializer() {
        return exceptionSerializer;
    }

    /**
     * Whether this descriptor will auto generate ACLs for each call.
     */
    public boolean autoAcl() {
        return autoAcl;
    }

    /**
     * The manually configured ACLs for this service.
     */
    public PSequence<ServiceAcl> acls() {
        return acls;
    }

    public HeaderTransformer protocolNegotiationStrategy() {
        return protocolNegotiationStrategy;
    }

    public HeaderTransformer serviceIdentificationStrategy() {
        return serviceIdentificationStrategy;
    }

    /**
     * Whether this is a locatable service.
     *
     * Locatable services are registered with the service locator and/or gateway, so that they can be consumed by other
     * services.  Services that are not locatable are typically services for infrastructure purposes, such as providing
     * metrics.
     */
    public boolean locatableService() {
        return locatableService;
    }

    /**
     * Provide a custom path param serializer for the given path param type.
     *
     * @param pathParamType The path param type.
     * @param pathParamSerializer The path param serializer.
     * @return A copy of this descriptor.
     */
    public <T> Descriptor with(Class<T> pathParamType, PathParamSerializer<T> pathParamSerializer) {
        return with((Type) pathParamType, pathParamSerializer);
    }

    /**
     * Provide a custom path param serializer for the given path param type.
     *
     * @param pathParamType The path param type.
     * @param pathParamSerializer The path param serializer.
     * @return A copy of this descriptor.
     */
    public Descriptor with(Type pathParamType, PathParamSerializer<?> pathParamSerializer) {
        return replaceAllPathParamSerializers(pathParamSerializers.plus(pathParamType, pathParamSerializer));
    }

    /**
     * Provide a custom MessageSerializer for the given message type.
     *
     * @param messageType The type of the message.
     * @param messageSerializer The message serializer for that type.
     * @return A copy of this descriptor.
     */
    public <T> Descriptor with(Class<T> messageType, MessageSerializer<T, ?> messageSerializer) {
        return with((Type) messageType, messageSerializer);
    }

    /**
     * Provide a custom MessageSerializer for the given message type.
     *
     * @param messageType The type of the message.
     * @param messageSerializer The message serializer for that type.
     * @return A copy of this descriptor.
     */
    public Descriptor with(Type messageType, MessageSerializer<?, ?> messageSerializer) {
        return replaceAllMessageSerializers(messageSerializers.plus(messageType, messageSerializer));
    }

    /**
     * Add the given service calls to this service.
     *
     * @param calls The calls to add.
     * @return A copy of this descriptor with the new calls added.
     */
    public Descriptor with(Call<?, ?>... calls) {
        return replaceAllCalls(this.calls.plusAll(Arrays.asList(calls)));
    }

    /**
     * Replace all the service calls provided by this descriptor with the the given service calls.
     *
     * @param calls The calls to replace the existing ones with.
     * @return A copy of this descriptor with the new calls.
     */
    public Descriptor replaceAllCalls(PSequence<Call<?, ?>> calls) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    /**
     * Replace all the path param serializers registered with this descriptor with the the given path param serializers.
     *
     * @param pathParamSerializers The path param serializers to replace the existing ones with.
     * @return A copy of this descriptor with the new path param serializers.
     */
    public Descriptor replaceAllPathParamSerializers(PMap<Type, PathParamSerializer<?>> pathParamSerializers) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    /**
     * Replace all the message serializers registered with this descriptor with the the given message serializers.
     *
     * @param messageSerializers The message serializers to replace the existing ones with.
     * @return A copy of this descriptor with the new message serializers.
     */
    public Descriptor replaceAllMessageSerializers(PMap<Type, MessageSerializer<?, ?>> messageSerializers) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    /**
     * Use the given exception serializer to serialize and deserialized exceptions handled by this service.
     *
     * @param exceptionSerializer The exception handler to use.
     * @return A copy of this descriptor.
     */
    public Descriptor with(ExceptionSerializer exceptionSerializer) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    /**
     * Use the given serializer factory with this service.
     *
     * @param serializerFactory The serializer factory to use.
     * @return A copy of this descriptor.
     */
    public Descriptor with(SerializerFactory serializerFactory) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    /**
     * Set whether the service calls in this descriptor should default to having an ACL automatically generated for
     * them.
     *
     * By default, this will not happen.
     *
     * Note that each service call can override this by calling withAutoAcl on them.
     *
     * @param autoAcl Whether autoAcl should be true.
     * @return A copy of this descriptor.
     */
    public Descriptor withAutoAcl(boolean autoAcl) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }


    /**
     * Add the given manual ACLs.
     *
     * If auto ACLs are configured, these will be added in addition to the auto ACLs.
     *
     * @param acls The ACLs to add.
     * @return A copy of this descriptor.
     */
    public Descriptor with(ServiceAcl... acls) {
        return replaceAllAcls(this.acls.plusAll(Arrays.asList(acls)));
    }

    /**
     * Replace all the ACLs with the given ACL sequence.
     *
     * This will not replace ACLs generated by autoAcl, to disable autoAcl, turn it off.
     *
     * @param acls The ACLs to use.
     * @return A copy of this descriptor.
     */
    public Descriptor replaceAllAcls(PSequence<ServiceAcl> acls) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    public Descriptor withProtocolNegotiationStrategy(HeaderTransformer protocolNegotiationStrategy) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    public Descriptor withServiceIdentificationStrategy(HeaderTransformer serviceIdentificationStrategy) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }

    /**
     * Set whether this service is locatable.
     *
     * Locatable services are registered with the service locator and/or gateway, so that they can be consumed by other
     * services.  Services that are not locatable are typically services for infrastructure purposes, such as providing
     * metrics.
     *
     * @param locatableService Whether this service should be locatable or not.
     * @return A copy of this descriptor.
     */
    public Descriptor withLocatableService(boolean locatableService) {
        return new Descriptor(name, calls, pathParamSerializers, messageSerializers, serializerFactory, exceptionSerializer, autoAcl, acls, protocolNegotiationStrategy, serviceIdentificationStrategy, locatableService);
    }


}

File: service/scaladsl/api/src/main/scala/com/lightbend/lagom/scaladsl/api/Descriptor.scala
/*
 * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
 */
package com.lightbend.lagom.scaladsl.api

import akka.NotUsed
import akka.util.ByteString
import com.lightbend.lagom.scaladsl.api.broker.Topic.TopicId
import com.lightbend.lagom.scaladsl.api.deser.{ DefaultExceptionSerializer, ExceptionSerializer, MessageSerializer }
import com.lightbend.lagom.scaladsl.api.transport.{ HeaderFilter, Method, UserAgentHeaderFilter }

import scala.collection.immutable
import scala.reflect.ClassTag

/**
 * Describes a service.
 *
 * A descriptor is a set of call and topic descriptors that the service provides, coupled with metadata about how the
 * service and its calls are to be served. Metadata may include versioning and migrations, SLA's, sharding
 * hints, circuit breaker strategies etc.
 */
sealed trait Descriptor {
  import Descriptor._

  /**
   * The name of this service.
   *
   * This will be used for service lookups.
   */
  val name: String

  /**
   * The service calls that this service provides.
   */
  val calls: immutable.Seq[Call[_, _]]

  /**
   * The topics that this service publishes.
   */
  val topics: immutable.Seq[TopicCall[_]]

  /**
   * An exception serializer, used for handling exceptions on either end.
   */
  val exceptionSerializer: ExceptionSerializer

  /**
   * Whether this service should automatically have all its service calls made publicly routable by the service gateway.
   */
  val autoAcl: Boolean

  /**
   * The ACLs associated with this service.
   */
  val acls: immutable.Seq[ServiceAcl]

  /**
   * The header filter to apply across all service calls on this service
   */
  val headerFilter: HeaderFilter

  /**
   * Whether this service should be locatable by the service locator.
   */
  val locatableService: Boolean

  /**
   * The default circuit breaker to use for this service.
   */
  val circuitBreaker: CircuitBreaker

  def addCalls(calls: Call[_, _]*): Descriptor = withCalls(this.calls ++ calls: _*)
  def withCalls(calls: Call[_, _]*): Descriptor

  def addTopics(topics: TopicCall[_]*): Descriptor = withTopics(this.topics ++ topics: _*)
  def withTopics(topics: TopicCall[_]*): Descriptor

  def withExceptionSerializer(exceptionSerializer: ExceptionSerializer): Descriptor
  def withAutoAcl(autoAcl: Boolean): Descriptor

  def addAcls(acls: ServiceAcl*): Descriptor = withAcls(this.acls ++ acls: _*)
  def withAcls(acls: ServiceAcl*): Descriptor

  def withHeaderFilter(headerFilter: HeaderFilter): Descriptor
  def withLocatableService(locatableService: Boolean): Descriptor
  def withCircuitBreaker(circuitBreaker: CircuitBreaker): Descriptor
}

object Descriptor {

  def apply(name: String): Descriptor = {
    DescriptorImpl(name)
  }

  /**
   * Describes a service call.
   */
  sealed trait Call[Request, Response] {

    /**
     * The id of the call.
     *
     * Used for routing purposes, may be a name, or could be a REST method and path pattern.
     */
    val callId: CallId

    /**
     * A holder for the service call.
     *
     * This holds a reference to the service call, in an implementation specific way.
     */
    val serviceCallHolder: ServiceCallHolder

    /**
     * The request message serializer.
     */
    val requestSerializer: MessageSerializer[Request, _]

    /**
     * The response message serializer.
     */
    val responseSerializer: MessageSerializer[Response, _]

    /**
     * The configured circuit breaker.
     *
     * @return Some value if this service call wants to override the circuit breaker configured for its service,
     *         otherwise empty.
     */
    val circuitBreaker: Option[CircuitBreaker]

    /**
     * Whether this service call should automatically define an ACL for the router to route external calls to it.
     *
     * @return Some value if this service call explicitly decides that it should have an auto ACL defined for it,
     *         otherwise empty.
     */
    val autoAcl: Option[Boolean]

    /**
     * Return a copy of this call with the given service call holder configured.
     */
    def withServiceCallHolder(serviceCallHolder: ServiceCallHolder): Call[Request, Response]

    /**
     * Return a copy of this call with the given request serializer configured.
     */
    def withRequestSerializer(requestSerializer: MessageSerializer[Request, _]): Call[Request, Response]

    /**
     * Return a copy of this call with the given request serializer configured.
     */
    def withResponseSerializer(responseSerializer: MessageSerializer[Response, _]): Call[Request, Response]

    /**
     * Return a copy of this call with the given circuit breaker configured.
     *
     * This will override the circuit breaker configured on the service descriptor.
     */
    def withCircuitBreaker(circuitBreaker: CircuitBreaker): Call[Request, Response]

    /**
     * Return a copy of this call with autoAcl configured.
     *
     * This will override auto ACL setting configured on the service descriptor.
     */
    def withAutoAcl(autoAcl: Boolean): Call[Request, Response]
  }

  val NoCall: Call[NotUsed, NotUsed] = CallImpl(NamedCallId("NONE"), new ServiceCallHolder {},
    MessageSerializer.NotUsedMessageSerializer, MessageSerializer.NotUsedMessageSerializer, None, None)

  /**
   * A call identifier.
   *
   * Call identifiers can be a simple name, or can be more complex identifiers that use REST methods and paths to
   * extract information and correspondingly route calls.
   */
  sealed trait CallId

  /**
   * A named call identifier.
   */
  sealed trait NamedCallId extends CallId {

    /**
     * The name of the call.
     */
    val name: String
  }

  object NamedCallId {

    /**
     * Create a named call identifier with the given name.
     */
    def apply(name: String): NamedCallId = NamedCallIdImpl(name)
  }

  /**
   * A path based call identifier.
   */
  sealed trait PathCallId extends CallId {

    /**
     * The pattern for path.
     */
    val pathPattern: String
  }

  object PathCallId {

    /**
     * Create a path based call identifier with the given path pattern.
     */
    def apply(pathPattern: String): PathCallId = PathCallIdImpl(pathPattern)
  }

  /**
   * A REST/HTTP call identifier.
   */
  sealed trait RestCallId extends CallId {

    /**
     * The HTTP method for the call.
     *
     * The method will only be used for strict REST calls. For other calls, such as calls implemented by WebSockets
     * or other transports, the method may be ignored completely.
     */
    val method: Method

    /**
     * The pattern for the path.
     */
    val pathPattern: String
  }

  object RestCallId {

    /**
     * Create a REST call identifier with the given method and path pattern.
     */
    def apply(method: Method, pathPattern: String): RestCallId = RestCallIdImpl(method, pathPattern)
  }

  /**
   * Holds the service call itself.
   *
   * The implementations of this are intentionally opaque, as the mechanics of how the service call implementation
   * gets passed around is internal to Lagom.
   */
  trait ServiceCallHolder

  /**
   * Describes a message broker topic.
   */
  sealed trait TopicCall[Message] {

    /**
     * The identifier for teh topic.
     */
    val topicId: TopicId

    /**
     * Holds the topic implementation.
     */
    val topicHolder: TopicHolder

    /**
     * The serializer for the topics messages.
     */
    val messageSerializer: MessageSerializer[Message, ByteString]

    /**
     * Some implementation specific properties for the topic.
     */
    val properties: Properties[Message]

    /**
     * Return a copy of this topic call with the given topic holder configured.
     */
    def withTopicHolder(topicHolder: TopicHolder): TopicCall[Message]

    /**
     * Add a property to this topic call.
     */
    def addProperty[T](property: Property[Message, T], value: T): TopicCall[Message]
  }

  /**
   * Holds the topic implementation.
   *
   * The implementations of this are intentionally opaque, as the mechanics of how the service call implementation
   * gets passed around is internal to Lagom.
   */
  trait TopicHolder

  /**
   * Properties of a topic call.
   */
  sealed trait Properties[Message] {
    /**
     * Get the given property.
     */
    def get[T](property: Property[Message, T]): Option[T]

    /**
     * Add the given property.
     */
    def +[T](propertyValue: (Property[Message, T], T)): Properties[Message]
  }

  object Properties {
    def empty[Message]: Properties[Message] = PropertiesImpl(Map())
  }

  /**
   * A property.
   */
  sealed trait Property[-Message, T] {
    /**
     * The class of the value.
     */
    val valueClass: Class[T]

    /**
     * The name of the property.
     */
    val name: String
  }

  object Property {
    def apply[Message, T: ClassTag](name: String): Property[Message, T] = {
      PropertyImpl(implicitly[ClassTag[T]].runtimeClass.asInstanceOf[Class[T]], name)
    }
  }

  private[api] case class DescriptorImpl(
    name:                String,
    calls:               immutable.Seq[Call[_, _]]   = Nil,
    topics:              immutable.Seq[TopicCall[_]] = Nil,
    exceptionSerializer: ExceptionSerializer         = DefaultExceptionSerializer.Unresolved,
    autoAcl:             Boolean                     = false,
    acls:                immutable.Seq[ServiceAcl]   = Nil,
    headerFilter:        HeaderFilter                = UserAgentHeaderFilter,
    locatableService:    Boolean                     = true,
    circuitBreaker:      CircuitBreaker              = CircuitBreaker.PerNode
  ) extends Descriptor {
    override def withCalls(calls: Call[_, _]*): Descriptor = copy(calls = calls.to[immutable.Seq])
    override def withTopics(topics: TopicCall[_]*): Descriptor = copy(topics = topics.to[immutable.Seq])
    override def withExceptionSerializer(exceptionSerializer: ExceptionSerializer): Descriptor = copy(exceptionSerializer = exceptionSerializer)
    override def withAutoAcl(autoAcl: Boolean): Descriptor = copy(autoAcl = autoAcl)
    override def withAcls(acls: ServiceAcl*): Descriptor = copy(acls = acls.to[immutable.Seq])
    override def withHeaderFilter(headerFilter: HeaderFilter): Descriptor = copy(headerFilter = headerFilter)
    override def withLocatableService(locatableService: Boolean): Descriptor = copy(locatableService = locatableService)
    override def withCircuitBreaker(circuitBreaker: CircuitBreaker): Descriptor = copy(circuitBreaker = circuitBreaker)
  }

  private[api] case class CallImpl[Request, Response](
    callId:             CallId,
    serviceCallHolder:  ServiceCallHolder,
    requestSerializer:  MessageSerializer[Request, _],
    responseSerializer: MessageSerializer[Response, _],
    circuitBreaker:     Option[CircuitBreaker]         = None,
    autoAcl:            Option[Boolean]                = None
  ) extends Call[Request, Response] {
    override def withServiceCallHolder(serviceCallHolder: ServiceCallHolder): Call[Request, Response] = copy(serviceCallHolder = serviceCallHolder)
    override def withRequestSerializer(requestSerializer: MessageSerializer[Request, _]): Call[Request, Response] = copy(requestSerializer = requestSerializer)
    override def withResponseSerializer(responseSerializer: MessageSerializer[Response, _]): Call[Request, Response] = copy(responseSerializer = responseSerializer)
    override def withCircuitBreaker(circuitBreaker: CircuitBreaker): Call[Request, Response] = copy(circuitBreaker = Some(circuitBreaker))
    override def withAutoAcl(autoAcl: Boolean): Call[Request, Response] = copy(autoAcl = Some(autoAcl))
  }

  private[api] case class NamedCallIdImpl(name: String) extends NamedCallId
  private[api] case class PathCallIdImpl(pathPattern: String) extends PathCallId
  private[api] case class RestCallIdImpl(method: Method, pathPattern: String) extends RestCallId

  private[api] case class TopicCallImpl[Message](
    topicId:           TopicId,
    topicHolder:       TopicHolder,
    messageSerializer: MessageSerializer[Message, ByteString],
    properties:        Properties[Message]                    = Properties.empty[Message]
  ) extends TopicCall[Message] {
    override def withTopicHolder(topicHolder: TopicHolder): TopicCall[Message] = copy(topicHolder = topicHolder)
    override def addProperty[T](property: Property[Message, T], value: T): TopicCall[Message] =
      copy(properties = properties + (property -> value))
  }

  private case class PropertyImpl[Message, T](valueClass: Class[T], name: String) extends Property[Message, T]
  private case class PropertiesImpl[Message](properties: Map[Property[Message, _], _]) extends Properties[Message] {
    override def get[T](property: Property[Message, T]): Option[T] = properties.get(property).asInstanceOf[Option[T]]
    override def +[T](propertyValue: (Property[Message, T], T)): Properties[Message] = PropertiesImpl(properties + propertyValue)
  }
}

sealed trait ServiceAcl {
  val method: Option[Method]
  val pathRegex: Option[String]
}

object ServiceAcl {
  def apply(method: Option[Method] = None, pathRegex: Option[String] = None): ServiceAcl = ServiceAclImpl(method, pathRegex)

  private case class ServiceAclImpl(method: Option[Method], pathRegex: Option[String]) extends ServiceAcl
}

sealed trait CircuitBreaker

object CircuitBreaker {
  case object None extends CircuitBreaker
  case object PerNode extends CircuitBreaker
  case object PerService extends CircuitBreaker

  sealed trait CircuitBreakerId extends CircuitBreaker {
    val id: String
  }

  def identifiedBy(id: String): CircuitBreaker = CircuitBreakerIdImpl(id)

  private case class CircuitBreakerIdImpl(id: String) extends CircuitBreakerId
}
